must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,207 @@
+ */
+public class MustCallAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+  /** The top annotation. */
+  final AnnotationMirror TOP;
+
+  /** The bottom annotation, which is the default in unannotated code. */
+  final AnnotationMirror BOTTOM;
+
+  /** The polymorphic qualifier */
+  final AnnotationMirror POLY;
+
+  /**
+   * Default constructor matching super. Should be called automatically.
+   *
+   * @param checker the checker associated with this type factory
+   */
+  public MustCallAnnotatedTypeFactory(final BaseTypeChecker checker) {
+    super(checker);
+    TOP = AnnotationBuilder.fromClass(elements, MustCallAny.class);
+    BOTTOM = createMustCall();
+    POLY = AnnotationBuilder.fromClass(elements, PolyMustCall.class);
+    this.postInit();
+  }
+
+  /** Treat non-owning method parameters as @MustCallAny to avoid false positives. */
+  @Override
+  public void methodFromUsePreSubstitution(ExpressionTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration;
+    if (tree instanceof MethodInvocationTree) {
+      declaration = TreeUtils.elementFromUse((MethodInvocationTree) tree);
+    } else if (tree instanceof MemberReferenceTree) {
+      declaration = (ExecutableElement) TreeUtils.elementFromTree(tree);
+    } else {
+      throw new BugInCF("unexpected type of method tree: " + tree.getKind());
+    }
+    for (int i = 0; i < type.getParameterTypes().size(); i++) {
+      Element paramDecl = declaration.getParameters().get(i);
+      if (getDeclAnnotation(paramDecl, NotOwning.class) != null) {
+        AnnotatedTypeMirror paramType = type.getParameterTypes().get(i);
+        paramType.replaceAnnotation(TOP);
+        // Descend into a varargs array
+        if (declaration.isVarArgs() && i == declaration.getParameters().size() - 1) {
+          AnnotatedTypeMirror componentType = ((AnnotatedArrayType) paramType).getComponentType();
+          componentType.replaceAnnotation(TOP);
+        }
+      }
+    }
+    super.methodFromUsePreSubstitution(tree, type);
+  }
+
+  /**
+   * Creates a {@link MustCall} annotation whose values are the given strings.
+   *
+   * @param val the methods that should be called
+   * @return an annotation indicating that the given methods should be called
+   */
+  public AnnotationMirror createMustCall(final String... val) {
+    AnnotationBuilder builder = new AnnotationBuilder(processingEnv, MustCall.class);
+    Arrays.sort(val);
+    builder.setValue("value", val);
+    return builder.build();
+  }
+
+  @Override
+  public QualifierHierarchy createQualifierHierarchy() {
+    return new MustCallQualifierHierarchy(this.getSupportedTypeQualifiers(), this.elements);
+  }
+
+  /**
+   * The qualifier hierarchy is responsible for lub, glb, and subtyping between qualifiers without
+   * declaratively-defined subtyping relationships, like our @MustCall annotation.
+   */
+  private class MustCallQualifierHierarchy extends ElementQualifierHierarchy {
```

Comment:
Please extend `MostlyNoElementQualifierHierarchy` instead.  That will enable you to simplify the implementation.